README.md
# Fourmolu

[![License BSD3](https://img.shields.io/badge/license-BSD3-brightgreen.svg)](http://opensource.org/licenses/BSD-3-Clause)
[![Hackage](https://img.shields.io/hackage/v/fourmolu.svg?style=flat)](https://hackage.haskell.org/package/fourmolu)

* [Configuration](#configuration)
* [Building and installation](#building-and-installation)
* [Usage](#usage)
    * [Editor integration](#editor-integration)
    * [Magic comments](#magic-comments)
    * [Exit codes](#exit-codes)
* [Limitations](#limitations)
* [Contributing](#contributing)
* [License](#license)

Fourmolu is a formatter for Haskell source code. It is a fork of [Ormolu](https://github.com/tweag/ormolu), with the intention to continue to merge upstream improvements.

We share all bar one of Ormolu's goals:

* Using GHC's own parser to avoid parsing problems caused by
  [`haskell-src-exts`](https://hackage.haskell.org/package/haskell-src-exts).
* Let some whitespace be programmable. The layout of the input influences
  the layout choices in the output. This means that the choices between
  single-line/multi-line layouts in certain situations are made by the user,
  not by an algorithm. This makes the implementation simpler and leaves some
  control to the user while still guaranteeing that the formatted code is
  stylistically consistent.
* Writing code in such a way so it's easy to modify and maintain.
* That formatting style aims to result in minimal diffs.
* Choose a style compatible with modern dialects of Haskell. As new Haskell
  extensions enter broad use, we may change the style to accommodate them.
* Idempotence: formatting already formatted code doesn't change it.
* Be well-tested and robust so that the formatter can be used in large
  projects.
* ~~Implementing one “true” formatting style which admits no configuration.~~ We allow configuration of various parameters, via CLI options or config files. We encourage any contributions which add further flexibility.

## Configuration

Fourmolu looks for a `fourmolu.yaml` file in all parents of the current directory, followed by [the XDG config directory](https://hackage.haskell.org/package/directory/docs/System-Directory.html#v:XdgConfig). A complete configuration file, corresponding to Fourmolu's default options, looks like:

```yaml
indentation: 4
comma-style: leading # for lists, tuples etc. - can also be 'trailing'
record-brace-space: false # rec {x = 1} vs. rec{x = 1}
indent-wheres: false # 'false' means save space by only half-indenting the 'where' keyword
diff-friendly-import-export: true # 'false' uses Ormolu-style lists
respectful: true # don't be too opinionated about newlines etc.
haddock-style: multi-line # '--' vs. '{-'
newlines-between-decls: 1 # number of newlines between top-level declarations
```

See [here](fourmolu.yaml) for a config to simulate the behaviour of Ormolu.

These options can also be set on the command line (which takes precedence over config files). Run `fourmolu -h` to see all options.

## Building and installation

Simply run `cabal v2-install fourmolu`, to install the latest release from Hackage.

You can also clone this repository, then build with Cabal or Stack.

## Usage

The following will print the formatted output to the standard output.

```console
$ fourmolu Module.hs
```

Add `-i` (or `--mode inplace`) to replace the contents of the input file with the formatted output.

```console
$ fourmolu -i Module.hs
```

Use `find` to format a tree recursively:

```console
$ fourmolu -i $(find . -name '*.hs')
```

Or find all files in a project with `git ls-files`:

```console
$ fourmoluormolu --mode inplace $(git ls-files '*.hs')
```

To check if files are are already formatted (useful on CI):

```console
$ fourmoluormolu --mode check $(find . -name '*.hs')
```

### Editor integration

Fourmolu can be integrated with your editor via the [Haskell Language Server](https://github.com/haskell/haskell-language-server).

### Magic comments

Fourmolu understands two magic comments:

```haskell
{- FOURMOLU_DISABLE -}
```

and

```haskell
{- FOURMOLU_ENABLE -}
```

This allows us to disable formatting selectively for code between these
markers or disable it for the entire file. To achieve the latter, just put
`{- FOURMOLU_DISABLE -}` at the very top. Note that for Fourmolu to work the
source code must still be parseable even when the disabled regions are
omitted. Because of that the magic comments cannot be placed arbitrarily,
but rather must enclose independent top-level definitions.

`{- ORMOLU_DISABLE -}` and `{- ORMOLU_ENABLE -}`, respectively, can be used to the same effect,
and the two styles of magic comments can be mixed.

### Exit codes

Exit code | Meaning
----------|-----------------------------------------------
0         | Success
1         | General problem
2         | CPP used (deprecated)
3         | Parsing of original input failed
4         | Parsing of formatted code failed
5         | AST of original and formatted code differs
6         | Formatting is not idempotent
7         | Unrecognized GHC options
100       | In checking mode: unformatted files
101       | Inplace and check modes do not work with stdin
102       | Other issue (with multiple input files)
400       | Failed to load Fourmolu configuration file

## Limitations

* CPP support is experimental. CPP is virtually impossible to handle
  correctly, so we process them as a sort of unchangeable snippets. This
  works only in simple cases when CPP conditionals surround top-level
  declarations. See the [CPP](https://github.com/tweag/ormolu/blob/master/DESIGN.md#cpp) section in the design notes for a
  discussion of the dangers.
* Input modules should be parsable by Haddock, which is a bit stricter
  criterion than just being valid Haskell modules.
* Various minor idempotence issues, most of them are related to comments.
* Fourmolu is in a fairly early stage of development. The implementation should be as stable as Ormolu, as it only makes minimal changes, and is extensively tested. But the default configuration style may change in some minor ways in the near future, as we make more options available. It will always be possible to replicate the old default behaviour with a suitable `fourmolu.yaml`.

## Contributing

If there are any options you'd like to see, let us know. If it's not too complicated to implement (and especially if you implement it yourself!) then we'll probably add it.

Run `cabal test` and `./format.sh` before submitting any pull requests.

## License

See [LICENSE.md](LICENSE.md).

Copyright © 2018–2020 Tweag I/O, 2020-present Matt Parsons

## Acknowledgements

The vast majority of work here has been done by the Ormolu developers, and thus they deserve almost all of the credit. This project is simply intended as a haven for those of us who admire their work, but can't quite get on board with some of their decisions when it comes down to the details.


CHANGELOG.md
## Unreleased

* Fixed issue with multiline haddocks in indented contexts ([#64](https://github.com/fourmolu/fourmolu/pull/64))
* Avoid renaming magic comments ([#90](https://github.com/fourmolu/fourmolu/pull/90))

### Upstream changes:

#### Unreleased

* Data declarations with multiline kind signatures are now formatted
  correctly. [Issue 749](https://github.com/tweag/ormolu/issues/749).

#### Ormolu 0.2.0.0

* Now standalone kind signatures are grouped with type synonyms. [Issue
  683](https://github.com/tweag/ormolu/issues/683).

* `TemplateHaskell` is not enabled by default anymore. [Issue
  699](https://github.com/tweag/ormolu/issues/699).

* Made record dot pre-processor rendering idempotent in certain specific
  cases. [Issue 701](https://github.com/tweag/ormolu/issues/701).

* Added support for arrow command application. [Issue
  716](https://github.com/tweag/ormolu/issues/716).

* Switched to `ghc-lib-parser-9.0.1`. [PR
  722](https://github.com/tweag/ormolu/pull/722).
   * Support for the new language extensions:
      * `LexicalNegation`, `LinearTypes`: disabled by default
      * `QualifiedDo`: enabled by default
   * Due to [upstream changes in whitespace sensitity](
     https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.0#whitespace-sensitive-and-),
     `TypeApplications` is now *enabled* by default. [Issue
     452](https://github.com/tweag/ormolu/issues/452).
   * Haddocks on declarations in files without a `module` header are no longer
     deleted. [Issue 480](https://github.com/tweag/ormolu/issues/480).
   * Due to a change in Haddock parsing, empty Haddock comments on function
     arguments now get deleted.

* CTYPE pragmas are now preserved. [Issue 689](
  https://github.com/tweag/ormolu/issues/689).

* `TypeApplications` in data/type family instances are now supported. [Issue
  698](https://github.com/tweag/ormolu/issues/698).

* Formatting infix arrow command formations now preserves the AST. [Issue
  718](https://github.com/tweag/ormolu/issues/718).

* Surround code in brackets with spaces if it contains a `StarIsType` `*` to
  prevent unparseable output. [Issue 704](https://github.com/tweag/ormolu/issues/704).

* Formatting applied multiline constructs in do blocks now preserves the AST.
  [Issue 707](https://github.com/tweag/ormolu/issues/707).

  This will sometimes result in odd indentations, e.g. this snippet is a
  fixed point:
  ```haskell
  foo = do
    do
      (+1)
     1
  ```

* GHC options passed in via the CLI can now be overridden in local files.
  Previously, if an extension was disabled via the CLI, it could not be
  re-enabled per file.

* `NegativeLiterals` is no longer enabled by default. Also, spaces after
  negation via `-` are removed where possible. [Issue
  694](https://github.com/tweag/ormolu/issues/694).

* Minus signs in literal patterns are now preserved in all cases. [Issue
  733](https://github.com/tweag/ormolu/issues/733).

* Added support for left-to-right arrow application. [Issue
  737](https://github.com/tweag/ormolu/issues/737).

* Now `--mode check` fails on missing trailing blank lines. [Issue
  743](https://github.com/tweag/ormolu/issues/743).

* Fixed indentation of arrow forms in do blocks. [Issue
  739](https://github.com/tweag/ormolu/issues/739).

#### Ormolu 0.1.4.1

* Added command line option `--color` to control how diffs are printed.
  Standardized the way errors are printed.

#### Ormolu 0.1.4.0

* Added support for monad comprehensions. [Issue
  665](https://github.com/tweag/ormolu/issues/665).

* Fixed a bug when a space was inserted in front of promoted types even when
  it wasn't strictly necessary. [Issue
  668](https://github.com/tweag/ormolu/issues/668).

* Now the checking mode displays diffs per file when unformatted files are
  found. The rendering of the diffs is also improved. [Issue
  656](https://github.com/tweag/ormolu/issues/656).

#### Ormolu 0.1.3.1

* Fixed a problem with multiline record updates using the record dot
  preprocessor. [Issue 658](https://github.com/tweag/ormolu/issues/658).

## Fourmolu 0.3.0.0

* New config option `newlines-between-decls`, to choose the number of blank lines between top-level declarations.
* Minor CLI improvements. In particular, the set of valid values for each option is communicated more consistently.

## Fourmolu 0.2.0.0

* More consistent indentation. Previously, with indentation set to n, some constructs such as nested lists and tuples would use an ugly mix of n-space and 2-space indentation.

* New configuration options (see README or Hackage docs for details):
  * `comma-style`
  * `record-brace-space`
  * `indent-wheres`
  * `diff-friendly-import-export`
  * `respectful`
  * `haddock-style`

### Upstream changes:

#### Ormolu 0.1.3.0:

* Ormolu no longer overwrites already formatted files. [PR
  649](https://github.com/tweag/ormolu/pull/649).

* Now a space is guaranteed before ticked promoted types. [Issue
  631](https://github.com/tweag/ormolu/issues/631).

* Made formatting of single-line explicitly bidirectional pattern synonyms
  idempotent. [Issue 630](https://github.com/tweag/ormolu/issues/630).

#### Ormolu 0.1.2.0:

* Fixed the bug when comments in different styles got glued together after
  formatting. [Issue 589](https://github.com/tweag/ormolu/issues/589).

* Added `-i` as a shortcut for `--mode inplace`. [Issue
  467](https://github.com/tweag/ormolu/issues/467).

* Improved grouping of top-level declarations. [Issue
  466](https://github.com/tweag/ormolu/issues/466).

## Fourmolu 0.1.1.0

### Upstream changes:

#### Ormolu 0.1.1.0:

* Imports in a import lists are now normalized: duplicate imports are
  combined/eliminated intelligently.

* Import declarations that can be merged are now automatically merged.
  [Issue 414](https://github.com/tweag/ormolu/issues/414).

* The magic comments for disabling and enabling Ormolu now can encompass any
  fragment of code provided that the remaining code after exclusion of the

app/Main.hs
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE ApplicativeDo #-}
{-# LANGUAGE CPP #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}

module Main (main) where

import Control.Monad
import Data.Bool (bool)
import Data.Functor.Identity (Identity (..))
import Data.List (intercalate, sort)
import Data.Maybe (fromMaybe, mapMaybe)
import qualified Data.Text.IO as TIO
import Data.Version (showVersion)
import Development.GitRev
import Options.Applicative
import Ormolu
import Ormolu.Config
import Ormolu.Diff.Text (diffText, printTextDiff)
import Ormolu.Parser (manualExts)
import Ormolu.Terminal
import Ormolu.Utils (showOutputable)
import Paths_fourmolu (version)
import System.Directory (getCurrentDirectory)
import System.Exit (ExitCode (..), exitWith)
import qualified System.FilePath as FP
import System.IO (hPutStrLn, stderr)

-- | Entry point of the program.
main :: IO ()
main = do
  opts@Opts {..} <- execParser optsParserInfo
  let formatStdIn = do
        cur <- getCurrentDirectory
        cfg <- mkConfig cur opts
        formatOne optMode cfg Nothing
  exitCode <- case optInputFiles of
    [] -> formatStdIn
    ["-"] -> formatStdIn
    [x] -> flip (formatOne optMode) (Just x) =<< mkConfig x opts
    xs@(x : _) -> do
      cfg <- mkConfig x opts
      let selectFailure = \case
            ExitSuccess -> Nothing
            ExitFailure n -> Just n
      errorCodes <-
        mapMaybe selectFailure <$> mapM (formatOne optMode cfg . Just) (sort xs)
      return $
        if null errorCodes
          then ExitSuccess
          else
            ExitFailure $
              if all (== 100) errorCodes
                then 100
                else 102
  exitWith exitCode

-- | Format a single input.
formatOne ::
  -- | Mode of operation
  Mode ->
  -- | Configuration
  Config RegionIndices ->
  -- | File to format or stdin as 'Nothing'
  Maybe FilePath ->
  IO ExitCode
formatOne mode config mpath = withPrettyOrmoluExceptions (cfgColorMode config) $
  case FP.normalise <$> mpath of
    Nothing -> do
      r <- ormoluStdin config
      case mode of
        Stdout -> do
          TIO.putStr r
          return ExitSuccess
        _ -> do
          hPutStrLn
            stderr
            "This feature is not supported when input comes from stdin."
          -- 101 is different from all the other exit codes we already use.
          return (ExitFailure 101)
    Just inputFile -> do
      originalInput <- TIO.readFile inputFile
      formattedInput <- ormoluFile config inputFile
      case mode of
        Stdout -> do
          TIO.putStr formattedInput
          return ExitSuccess
        InPlace -> do
          -- Only write when the contents have changed, in order to avoid
          -- updating the modified timestamp if the file was already correctly
          -- formatted.
          when (formattedInput /= originalInput) $
            TIO.writeFile inputFile formattedInput
          return ExitSuccess
        Check ->
          case diffText originalInput formattedInput inputFile of
            Nothing -> return ExitSuccess
            Just diff -> do
              runTerm (printTextDiff diff) (cfgColorMode config) stderr
              -- 100 is different to all the other exit code that are emitted
              -- either from an 'OrmoluException' or from 'error' and
              -- 'notImplemented'.
              return (ExitFailure 100)

----------------------------------------------------------------------------
-- Command line options parsing

data Opts = Opts
  { -- | Mode of operation
    optMode :: !Mode,
    -- | Ormolu 'Config'
    optConfig :: !(Config RegionIndices),
    -- | Fourmolu-specific options
    optPrinterOpts :: !PrinterOptsPartial,
    -- | Haskell source files to format or stdin (when the list is empty)
    optInputFiles :: ![FilePath]
  }

-- | Mode of operation.
data Mode
  = -- | Output formatted source code to stdout
    Stdout
  | -- | Overwrite original file
    InPlace
  | -- | Exit with non-zero status code if
    -- source is not already formatted
    Check
  deriving (Eq, Show, Bounded, Enum)

optsParserInfo :: ParserInfo Opts
optsParserInfo =
  info (helper <*> ver <*> exts <*> optsParser) . mconcat $
    [fullDesc]
  where
    ver :: Parser (a -> a)
    ver =
      infoOption verStr . mconcat $
        [ long "version",
          short 'v',
          help "Print version of the program"
        ]
    verStr =
      intercalate
        "\n"
        [ unwords
            [ "fourmolu",
              showVersion version,
              $gitBranch,
              $gitHash
            ],
          "using ghc-lib-parser " ++ VERSION_ghc_lib_parser
        ]
    exts :: Parser (a -> a)
    exts =
      infoOption displayExts . mconcat $
        [ long "manual-exts",
          help "Display extensions that need to be enabled manually"
        ]
    displayExts = unlines $ sort (showOutputable <$> manualExts)

optsParser :: Parser Opts
optsParser =
  Opts
    <$> ( (fmap (bool Stdout InPlace) . switch . mconcat)
            [ short 'i',
              help "A shortcut for --mode inplace"
            ]
            <|> (option parseBoundedEnum . mconcat)
              [ long "mode",
                short 'm',
                metavar "MODE",
                value Stdout,
                help "Mode of operation: 'stdout' (the default), 'inplace', or 'check'"
              ]
        )
    <*> configParser
    <*> printerOptsParser
    <*> (many . strArgument . mconcat)
      [ metavar "FILE",
        help "Haskell source files to format or stdin (the default)"
      ]

configParser :: Parser (Config RegionIndices)
configParser =
  Config
    <$> (fmap (fmap DynOption) . many . strOption . mconcat)
      [ long "ghc-opt",
        short 'o',
        metavar "OPT",
        help "GHC options to enable (e.g. language extensions)"
      ]
    <*> (switch . mconcat)
      [ long "unsafe",
        short 'u',
        help "Do formatting faster but without automatic detection of defects"
      ]
    <*> (switch . mconcat)
      [ long "debug",
        short 'd',
        help "Output information useful for debugging"
      ]
    <*> (switch . mconcat)
      [ long "check-idempotence",
        short 'c',
        help "Fail if formatting is not idempotent"
      ]
    <*> (option parseBoundedEnum . mconcat)
      [ long "color",
        metavar "WHEN",
        value Auto,
        help "Colorize the output; WHEN can be 'never', 'always', or 'auto' (the default)"
      ]
    <*> ( RegionIndices
            <$> (optional . option auto . mconcat)
              [ long "start-line",
                metavar "START",
                help "Start line of the region to format (starts from 1)"
              ]
            <*> (optional . option auto . mconcat)
              [ long "end-line",
                metavar "END",
                help "End line of the region to format (inclusive)"
              ]
        )
    <*> pure defaultPrinterOpts

printerOptsParser :: Parser PrinterOptsPartial
printerOptsParser = do
  poIndentation <-
    (optional . option auto . mconcat)
      [ long "indentation",
        metavar "WIDTH",
        help $
          "Number of spaces per indentation step"
            <> showDefaultValue poIndentation
      ]
  poCommaStyle <-
    (optional . option parseBoundedEnum . mconcat)
      [ long "comma-style",
        metavar "STYLE",
        help $
          "How to place commas in multi-line lists, records etc: "
            <> showAllValues @CommaStyle
            <> showDefaultValue poCommaStyle
      ]
  poIndentWheres <-
    (optional . option parseBoundedEnum . mconcat)
      [ long "indent-wheres",
        metavar "BOOL",
        help $
          "Whether to indent 'where' bindings past the preceding body"
            <> " (rather than half-indenting the 'where' keyword)"
            <> showDefaultValue poIndentWheres
      ]
  poRecordBraceSpace <-
    (optional . option parseBoundedEnum . mconcat)
      [ long "record-brace-space",
        metavar "BOOL",
        help $
          "Whether to leave a space before an opening record brace"
            <> showDefaultValue poRecordBraceSpace
      ]
  poDiffFriendlyImportExport <-
    (optional . option parseBoundedEnum . mconcat)
      [ long "diff-friendly-import-export",
        metavar "BOOL",
        help $
          "Whether to make use of extra commas in import/export lists"
            <> " (as opposed to Ormolu's style)"
            <> showDefaultValue poDiffFriendlyImportExport
      ]
  poRespectful <-
    (optional . option parseBoundedEnum . mconcat)
      [ long "respectful",
        metavar "BOOL",
        help $
          "Give the programmer more choice on where to insert blank lines"
            <> showDefaultValue poRespectful
      ]
  poHaddockStyle <-
    (optional . option parseBoundedEnum . mconcat)
      [ long "haddock-style",
        metavar "STYLE",
        help $
          "How to print Haddock comments: "
            <> showAllValues @HaddockPrintStyle
            <> showDefaultValue poHaddockStyle
      ]
  poNewlinesBetweenDecls <-
    (optional . option auto . mconcat)
      [ long "newlines-between-decls",
        metavar "HEIGHT",
        help $
          "Number of spaces between top-level declarations"
            <> showDefaultValue poNewlinesBetweenDecls
      ]
  pure PrinterOpts {..}

----------------------------------------------------------------------------
-- Helpers

-- | A standard parser of CLI option arguments, applicable to arguments that
-- have a finite (preferably small) number of possible values. (Basically an
-- inverse of 'toCLIArgument'.)
parseBoundedEnum ::
  forall a.
  (Enum a, Bounded a, ToCLIArgument a) =>
  ReadM a
parseBoundedEnum =
  eitherReader
    ( \s ->
        case lookup s argumentToValue of
          Just v -> Right v
          Nothing ->
            Left $
              "unknown value: '"
                <> s
                <> "'\nValid values are: "
                <> showAllValues @a
                <> "."
    )
  where
    argumentToValue = map (\x -> (toCLIArgument x, x)) [minBound ..]

-- | Values that appear as arguments of CLI options and thus have
-- a corresponding textual representation.
class ToCLIArgument a where
  -- | Convert a value to its representation as a CLI option argument.
  toCLIArgument :: a -> String

  -- | Convert a value to its representation as a CLI option argument wrapped
  -- in apostrophes.
  toCLIArgument' :: a -> String
  toCLIArgument' x = "'" <> toCLIArgument x <> "'"

instance ToCLIArgument Bool where
  toCLIArgument True = "true"
  toCLIArgument False = "false"

instance ToCLIArgument CommaStyle where
  toCLIArgument Leading = "leading"
  toCLIArgument Trailing = "trailing"

instance ToCLIArgument Int where
  toCLIArgument = show

instance ToCLIArgument HaddockPrintStyle where
  toCLIArgument HaddockSingleLine = "single-line"
  toCLIArgument HaddockMultiLine = "multi-line"

instance ToCLIArgument Mode where
  toCLIArgument Stdout = "stdout"
  toCLIArgument InPlace = "inplace"
  toCLIArgument Check = "check"

instance ToCLIArgument ColorMode where
  toCLIArgument Never = "never"
  toCLIArgument Always = "always"
  toCLIArgument Auto = "auto"

showAllValues :: forall a. (Enum a, Bounded a, ToCLIArgument a) => String
showAllValues = format (map toCLIArgument' [(minBound :: a) ..])
  where
    format [] = []
    format [x] = x
    format [x1, x2] = x1 <> " or " <> x2
    format (x : xs) = x <> ", " <> format xs

-- | CLI representation of the default value of an option, formatted for
-- inclusion in the help text.
showDefaultValue ::
  ToCLIArgument a =>
  (PrinterOptsTotal -> Identity a) ->
  String
showDefaultValue =
  (" (default " <>)
    . (<> ")")
    . toCLIArgument'
    . runIdentity
    . ($ defaultPrinterOpts)

-- | Build the full config, by adding 'PrinterOpts' from a file, if found.
mkConfig :: FilePath -> Opts -> IO (Config RegionIndices)
mkConfig path Opts {..} = do
  filePrinterOpts <-
    loadConfigFile path >>= \case
      ConfigLoaded f po -> do
        hPutStrLn stderr $ "Loaded config from: " <> f
        printDebug $ show po
        return $ Just po
      ConfigParseError f (_pos, err) -> do
        -- we ignore '_pos' due to the note on 'Data.YAML.Aeson.decode1'
        hPutStrLn stderr $
          unlines
            [ "Failed to load " <> f <> ":",
              "  " <> err
            ]
        exitWith $ ExitFailure 400
      ConfigNotFound searchDirs -> do
        printDebug
          . unlines
          $ ("No " ++ show configFileName ++ " found in any of:") :
          map ("  " ++) searchDirs
        return Nothing
  return $
    optConfig
      { cfgPrinterOpts =
          fillMissingPrinterOpts
            (optPrinterOpts <> fromMaybe mempty filePrinterOpts)
            (cfgPrinterOpts optConfig)
      }
  where
    printDebug = when (cfgDebug optConfig) . hPutStrLn stderr
